src/raft.c

```diff
@@ -1238,334 +1236,271 @@ RRStatus RedisRaftInit(RedisModuleCtx *ctx, RedisRaftCtx *rr, RedisRaftConfig *c
+    if ((err = raft_recv_appendentries(rr->raft, node, msg, &resp)) != 0) {
+        snprintf(buf, sizeof(buf), "operation failed, error %d", err);
+        RedisModule_ReplyWithError(ctx, buf);
+        return;
     }
 
-    RedisModule_ReplyWithArray(req->ctx, 4);
-    RedisModule_ReplyWithLongLong(req->ctx, response.term);
-    RedisModule_ReplyWithLongLong(req->ctx, response.success);
-    RedisModule_ReplyWithLongLong(req->ctx, response.current_idx);
-    RedisModule_ReplyWithLongLong(req->ctx, response.msg_id);
-
-exit:
-    RaftReqFree(req);
+    RedisModule_ReplyWithArray(ctx, 4);
+    RedisModule_ReplyWithLongLong(ctx, resp.term);
+    RedisModule_ReplyWithLongLong(ctx, resp.success);
+    RedisModule_ReplyWithLongLong(ctx, resp.current_idx);
+    RedisModule_ReplyWithLongLong(ctx, resp.msg_id);
 }
 
-void handleCfgChange(RedisRaftCtx *rr, RaftReq *req)
+void handleCfgAdd(RedisRaftCtx *rr,
+                  RedisModuleCtx *ctx, raft_node_id_t id, NodeAddr *addr)
 {
     raft_entry_t *entry;
     msg_entry_response_t response;
     int e;
 
-    if (checkRaftState(rr, req) == RR_ERROR ||
-        checkLeader(rr, req, NULL) == RR_ERROR) {
-        goto exit;
+    if (checkRaftState(rr, ctx) == RR_ERROR ||
+        checkLeader(rr, ctx, NULL) == RR_ERROR) {
+        return;
     }
 
-    short type;
+    if (hasNodeIdBeenUsed(rr, id)) {
+        RedisModule_ReplyWithError(ctx, "node id has already been used in this cluster");
+        return;
+    }
 
-    switch (req->type) {
-        case RR_CFGCHANGE_ADDNODE:
-            if (hasNodeIdBeenUsed(rr, req->r.cfgchange.id)) {
-                RedisModule_ReplyWithError(req->ctx,
-                               "node id has already been used in this cluster");
-                goto exit;
-            }
+    if (!id) {
+        id = makeRandomNodeId(rr);
+    }
 
-            type = RAFT_LOGTYPE_ADD_NONVOTING_NODE;
-            if (!req->r.cfgchange.id) {
-                req->r.cfgchange.id = makeRandomNodeId(rr);
-            }
-            break;
-        case RR_CFGCHANGE_REMOVENODE:
-            /* Validate it exists */
-            if (!raft_get_node(rr->raft, req->r.cfgchange.id)) {
-                RedisModule_ReplyWithError(req->ctx, "node id does not exist");
-                goto exit;
-            }
+    RaftCfgChange cfg = {
+        .id = id,
+        .addr = *addr
+    };
 
-            type = RAFT_LOGTYPE_REMOVE_NODE;
-            break;
-        default:
-            RedisModule_Assert(0);
+    entry = raft_entry_new(sizeof(cfg));
+    entry->id = rand();
+    entry->type = RAFT_LOGTYPE_ADD_NONVOTING_NODE;
+    memcpy(entry->data, &cfg, sizeof(cfg));
+
+    e = raft_recv_entry(rr->raft, entry, &response);
+    if (e != 0) {
+        raft_entry_release(entry);
+        replyRaftError(ctx, e);
+        return;
+    }
+
+    raft_entry_release(entry);
+
+    /* We don't have to block on add node, it's all through join which blocks itself */
```

Comment:
Don't we have to handle this like any consensus operation? If we're generating the reply now, how can we be sure we don't end up unblocking the client and returning it even if not committed?

Reply:
This is existing code, I didn't change, just moved. I agree, it makes to block here but probably we should handle as part of another PR.

Reply:
Opened an issue: https://github.com/RedisLabs/redisraft/issues/265